File: one-unity/core/development/common/game-account/Runtime/Scripts/Util/HttpClientApi.cs
namespace TPFive.Game.Account
{
    using System;
    using System.Collections.Generic;
    using System.Net.Http;
    using System.Net.Http.Headers;
    using System.Text;
    using System.Threading.Tasks;
    using UnityEngine;

    public class HttpClientApi
    {
        protected static readonly HttpClient Client = new HttpClient();

        private static readonly string TAG = "[XRAccount][HttpClientApi]";
        private static readonly string TagOfSignInName = "email";

        /*
         * openid: Required (Indicates that the token is an OIDC token)
         * offline_access: Requests a refresh token
         * profile: Requests access to the user's profile, we need nickname
         * username: Requests access to the user's username
         * email: Requests access to the user's email and email_verified
         */
        private static readonly string OidcScope = "openid offline_access profile username email";

        public HttpClientApi(string domain, string clientId, string accountServer)
        {
            Domain = domain;
            ClientID = clientId;
            AccountServer = accountServer;
        }

        public string ClientID { get; private set; }

        public string Domain { get; private set; }

        public string AccountServer { get; private set; }

        public Task<HttpResponseMessage> HttpSignIn(string name, string password)
        {
            // URL
            string urlString = GetUrlString(Domain, AccountConst.AuthingTokenEndpoint);
            Debug.Log($"{TAG} urlString: {urlString}");

            // Body
            var formBody = new Dictionary<string, string>
            {
                { "client_id", ClientID },
                { "grant_type", "password" },
                { "password", password },
                { TagOfSignInName, name },
                { "scope", OidcScope },
            };
            var content = new FormUrlEncodedContent(formBody);

            // Header
            content.Headers.ContentType = new MediaTypeHeaderValue("application/x-www-form-urlencoded");
            var request = new HttpRequestMessage(HttpMethod.Post, urlString)
            {
                Content = content,
            };
            return Client.SendAsync(request);
        }

        public Task<HttpResponseMessage> HttpRenewAuth(string refreshToken)
        {
            string urlString = GetUrlString(Domain, AccountConst.AuthingTokenEndpoint);
            Dictionary<string, string> body = new Dictionary<string, string>
            {
                { "client_id", ClientID },
                { "grant_type", "refresh_token" },
                { "refresh_token", refreshToken },
            };
            var content = new FormUrlEncodedContent(body);
            content.Headers.ContentType = new MediaTypeHeaderValue("application/x-www-form-urlencoded");
            var request = new HttpRequestMessage(HttpMethod.Post, urlString) { Content = content };
            return Client.SendAsync(request);
        }

        public Task<HttpResponseMessage> StartDeviceAuth()
        {
            string urlString = GetUrlString(AccountServer, AccountConst.DeviceAuthEndpoint);
            Dictionary<string, string> body = new Dictionary<string, string>
            {
                { "client_id", ClientID },
                { "scene", "APP_AUTH" },
            };
            var content = new FormUrlEncodedContent(body);
            var request = new HttpRequestMessage(HttpMethod.Post, urlString) { Content = content };
            return Client.SendAsync(request);
        }

        public Task<HttpResponseMessage> PollingDeviceCodeStatus(string userCode)
        {
            string urlString = GetUrlString(AccountServer, AccountConst.DeviceAuthStatusEndpoint, $"client_id={ClientID}&user_code={userCode}");
            var request = new HttpRequestMessage(HttpMethod.Get, urlString);
            return Client.SendAsync(request);
        }

        public Task<HttpResponseMessage> AuthingTokenToCredential(string authingToken)
        {
            string urlString = GetUrlString(Domain, AccountConst.DeviceAuthingIdTokenToAccessTokenEndpoint);
            Dictionary<string, string> body = new Dictionary<string, string>
            {
                { "client_id", ClientID },
                { "grant_type", AccountConst.IdTokenToCredentialsGrantType },
                { "redirect_uri", AccountConst.IdTokenToCredentialsRedirectUri },
                { "token", authingToken },
                { "scope", "openid profile email phone" },
            };

            var content = new FormUrlEncodedContent(body);
            var request = new HttpRequestMessage(HttpMethod.Post, urlString) { Content = content };
            return Client.SendAsync(request);
        }

        public Task<HttpResponseMessage> CreateGuestAccount(string nickname)
        {
            string urlString = GetUrlString(AccountServer, AccountConst.GuestCreateEndpoint);
            string jsonString = $"{{\"nickname\":\"{nickname}\"}}";
            Debug.Log($"{TAG} Request body = {jsonString}");
            var content = new StringContent(jsonString, Encoding.UTF8, "application/json");
            var request = new HttpRequestMessage(HttpMethod.Post, urlString) { Content = content };
            return Client.SendAsync(request);
        }

        private static string GetUrlString(string domain, string endpoints, string queryString = null)
        {
            Uri uri = new Uri(domain);
            return new UriBuilder
            {
                Scheme = uri.Scheme,
                Host = uri.Host,
                Port = uri.Port,
                Path = endpoints,
                Query = queryString,
            }.ToString();
        }
    }
}

File: one-mobile/flutter_project/lib/xrauth/util/http_client_api.dart
import 'dart:convert';

import 'package:http/http.dart' as http;
import 'package:tpfive/xrauth/constants/account_const.dart';
import 'package:tpfive/utils/logger.dart';

class HttpClientApi {
  static const String TAG = 'HttpClientApi';
  static const String _oidcScope =
      'openid offline_access profile username email';

  late String _authingDomain;
  late String _clientId;
  late String _serverDomain;

  HttpClientApi(String authingDomain, String clientId, String serverDomain) {
    _authingDomain = authingDomain;
    _clientId = clientId;
    _serverDomain = serverDomain;
  }

  Future<http.Response> signInByUsername(String username, String password) {
    Uri url = _getUri(_authingDomain, AccountConst.authingTokenEndpoint);
    Log.d(TAG, 'url: $url');
    return http.post(
      url,
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
      },
      body: {
        'client_id': _clientId,
        'grant_type': 'password',
        'email': username,
        'password': password,
        'scope': _oidcScope
      },
      encoding: Encoding.getByName('utf-8'),
    );
  }

  Future<http.Response> renewAuth(String refreshToken) {
    Uri url = _getUri(_authingDomain, AccountConst.authingTokenEndpoint);
    Log.d(TAG, 'url: $url.toString()');
    return http.post(
      url,
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
      },
      body: {
        'client_id': _clientId,
        'grant_type': 'refresh_token',
        'refresh_token': refreshToken
      },
      encoding: Encoding.getByName('utf-8'),
    );
  }

  Future<http.Response> startDeviceAuth() {
    Uri url = _getUri(_serverDomain, AccountConst.deviceAuthEndpoint);
    Log.d(TAG, 'url: $url.toString()');
    return http.post(
      url,
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
      },
      body: {'client_id': _clientId, 'scene': 'APP_AUTH'},
      encoding: Encoding.getByName('utf-8'),
    );
  }

  Future<http.Response> pollingDeviceCodeStatus(String userCode) {
    Uri url = _getUri(_serverDomain, AccountConst.deviceAuthStatusEndpoint,
        'client_id=$_clientId&user_code=$userCode');
    Log.d(TAG, 'url: $url.toString()');
    return http.get(url);
  }

  Future<http.Response> authingTokenToCredential(String authingToken) {
    Uri url = _getUri(
        _serverDomain, AccountConst.deviceAuthingIdTokenToAccessTokenEndpoint);
    Log.d(TAG, 'url: $url.toString()');
    return http.post(
      url,
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
      },
      body: {
        'client_id': _clientId,
        'grant_type': AccountConst.idTokenToCredentialsGrantType,
        'redirect_uri': AccountConst.idTokenToCredentialsRedirectUri,
        'token': authingToken,
        'scope': _oidcScope
      },
      encoding: Encoding.getByName('utf-8'),
    );
  }

  Future<http.Response> createGuestAccount(String nickname) {
    Uri uri = _getUri(_serverDomain, AccountConst.guestCreateEndpoint);
    Log.d(TAG, 'url: $uri.toString()');
    return http.post(
      uri,
      headers: {
        'Content-Type': 'application/json; charset=UTF-8',
      },
      body: json.encode({'nickname': nickname}),
      encoding: Encoding.getByName('utf-8'),
    );
  }

  Uri _getUri(String domain, String endpoints, [String? queryString]) {
    Uri uri = Uri.parse(domain);
    return Uri(
      scheme: uri.scheme,
      host: uri.host,
      port: uri.port,
      path: endpoints,
      query: queryString,
    );
  }

  bool isSuccessResponse(int httpCode) {
    return httpCode >= 200 && httpCode <= 299;
  }
}
